cljfmt/src/cljfmt/tool.clj

```diff
@@ -99,25 +100,77 @@
   ([a]   a)
   ([a b] (merge-with + a b)))
 
-(defn check-no-config
-  "The same as `check`, but ignores dotfile configuration."
-  [options]
-  (let [map*   (if (:parallel? options) pmap map)
-        counts (->> (:paths options)
-                    (mapcat (partial find-files options))
-                    (map* (partial check-one options))
-                    (map (fn [status]
-                           (print-file-status options status)
-                           (:counts status)))
-                    (reduce merge-counts))]
+(defn- merge-statuses
+  [a b]
+  (let [^File file (:file b)
+        path       (.getPath file)
+        diff       (:diff b)
+        exception  (:exception b)]
+    (cond-> a
+      true (update :counts #(merge-with + % (:counts b)))
+
+      (and file (= 1 (-> b :counts :incorrect)))
+      (assoc-in [:incorrect path] "has incorrect formatting")
+
+      (and file diff) (assoc-in [:incorrect path] diff)
+      (and file exception) (assoc-in [:error path] exception))))
+
+(defmulti report-cli (fn [event & _] event))
+(defmulti report-lib (fn [event & _] event))
+
+(defmethod report-cli :check/initial
+  [_ _ _]
+  nil)
+
+(defmethod report-cli :check/file
+  [_ context data]
+  (print-file-status context data)
+  (:counts data))
+
+(defmethod report-cli :check/summary
+  [_ summary _]
+  (let [counts (reduce merge-counts summary)]
     (print-final-count counts)
     (exit counts)))
 
+(defmethod report-lib :check/initial
+  [_ _ _]
+  nil)
+
+(defmethod report-lib :check/file
+  [_ _ data]
+  data)
+
+(defmethod report-lib :check/summary
+  [_ context _]
+  (reduce merge-statuses {} context))
+
+(defn check-no-config
+  "The same as `check`, but ignores dotfile configuration."
+  [{:keys [parallel? paths report] :as options
+    :or   {report report-cli}}]
+  (report :check/initial options nil)
+  (let [map*    (if parallel? pmap map)
+        summary (->> paths
+                     (mapcat (partial find-files options))
+                     (map* (partial check-one options))
+                     (map #(report :check/file options %)))]
+    (report :check/summary summary nil)))
+
 (defn check
   "Checks that the Clojure paths specified by the :paths option are
   correctly formatted."
   [options]
-  (let [opts (merge (config/load-config) options)]
+  (let [opts (merge {:report report-cli}
+                    (config/load-config) options)]
+    (check-no-config opts)))
+
+(defn check-paths
+  "Runs a check on the provided paths and returns a map detailing the results."
+  [paths & [options]]
+  (let [opts (merge {:report report-lib}
+                    (config/load-config) options
+                    {:paths paths})]
     (check-no-config opts)))
```

Comment:
Why add this function? Why not make it equivalent to using `report-lib` with `check`?

Reply:
Well, the entire point of this PR is to make this more ergonomic to use as a library. So this is the library entry point fn for checking paths.

Not sure I understand this:
> Why not make it equivalent to using report-lib with check?

Reply:
I mean something like:

```clojure
(check {:paths ["src"], :report report-lib})
```

Reply:
I will also note that this is a big part of why I started with a different namespace. Naming this fn in here is tricky in terms of revealing its intention (and so was naming the other namespace as you probably recall; so that's perhaps not a great solution either).

But the idea is you would call this fn when using cljfmt as a library to check paths (instead of strings), and the existing fn(s) are still for internal CLI usage. I'm open to ideas about how to better name this, document that intention, etc.

But I do think it makes sense to have a "library entrypoint" fn that doesn't require setting options you'll want every time to make it be a well-behaved library (e.g. not calling `System/exit`, not printing to stdout, etc.).

Reply:
I think I'd rather omit the `check-paths` and `fix-paths` wrapper functions for now. I don't think it's a huge deal to add an extra `:report` argument to cater for a relatively niche use-case.

I think it would also be a good idea to move `report-cli` to `cljfmt.report/console` and `report-lib` to `cljfmt.report/clojure`. However, don't feel you need to make that chage; I can always move things around after the PR is merged.

Reply:
OK, I can make both changes

Reply:
Would you prefer these changes as new, separate commits? Or should I just continue to amend the existing HEAD commit on this branch?

Reply:
I'll go ahead and push separate commits for easier review. Happy to squash later if desired.

Reply:
When I moved the reporting fns to the new `cljfmt.report` ns, it required one other change to avoid a circular dependency: The `project-path` fn is used by both `cljfmt.tool` fns and `cljfmt.report` fns, so it needed to be in a namespace they both required. `cljfmt.io` seemed like a natural fit, so I moved it there.